### PR TITLE
feat(debug): enhance torch codegen for paged attention support

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary
- Add tile broadcast ops, binary/unary expression dispatch delegates, view-based slicing, partial-tile padding, and FP32 matmul accumulation to `torch_codegen` — enabling end-to-end codegen of paged attention IR
- Add `check_shapes` parameter for optional runtime shape/dtype verification against IR type annotations
- Add system tests (`tests/st/codegen/`) comparing codegen output vs golden reference for static and dynamic paged attention

## Testing
- [x] 27 unit tests pass (`tests/ut/ir/printing/test_torch_codegen.py`)
- [x] 2 system tests pass (`tests/st/codegen/test_torch_codegen_paged_attention.py`)
- [x] Pre-commit hooks pass (ruff, pyright, headers)

### CODE REVIEW

File: python/pypto/debug/torch_codegen.py [hunk 1]
```diff
@@ -63,11 +63,23 @@ def _torch_dtype(dt: DataType) -> str:
 
 def _tile_load(tensor, offsets, shapes):
     slices = tuple(slice(o, o + s) for o, s in zip(offsets, shapes))
-    return tensor[slices].clone()
+    tile = tensor[slices].clone()
+    # Pad to requested shape if source is smaller (mirrors hardware valid_shapes)
+    if tile.shape != tuple(shapes):
+        padded = torch.zeros(shapes, dtype=tile.dtype)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/debug/torch_codegen.py]: Fixed in 3e1dfe46 — changed to `tile.new_zeros(shapes)` which preserves device and dtype.

File: python/pypto/debug/torch_codegen.py [hunk 2]
```diff
@@ -440,6 +454,10 @@ def visit_function(self, func: _ir.Function) -> None:
         params = [self._name_of(p) for p in func.params]
         self._emit(f"def {func.name}({', '.join(params)}):")
         self._indent += 1
+        if self._check_shapes:
+            for p in func.params:
+                # Params may receive partial tiles, so only check dtype.
+                self._emit_shape_dtype_check(self._name_of(p), p.type, shape=False)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/debug/torch_codegen.py]: Investigated — InCore kernel functions receive partial data for both TensorType and TileType params (boundary tiles). Skipping shape checks for all function params is intentional. Updated comment to clarify in 3e1dfe46.

File: python/pypto/debug/torch_codegen.py [hunk 3]
```diff
@@ -522,10 +540,64 @@ def visit_call(self, op: _ir.Call) -> None:
...
+        torch_dt = _torch_dtype(dtype)
+
+        self._emit(
+            f"assert isinstance({var_name}, torch.Tensor), "
+            f'f"Expected {var_name} to be a Tensor, got {{type({var_name}).__name__}}"'
+        )
+        if shape:
+            # Check if all dimensions are resolvable (ConstInt or locally-scoped
+            # Vars).  Dimensions from pl.dynamic() are Vars whose names are not
+            # in the generated scope, so we fall back to an ndim-only check.
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/debug/torch_codegen.py]: Fixed in 3e1dfe46 — updated comment to accurately describe the implementation.

File: python/pypto/debug/torch_codegen.py [hunk 4]
```diff
@@ -283,13 +296,13 @@ def _register_ops() -> None:
...
-    m["tile.matmul_acc"] = lambda a, _kw: f"({a[0]} + torch.matmul({a[1]}, {a[2]}))"
-    m["tile.matmul_bias"] = lambda a, _kw: f"(torch.matmul({a[0]}, {a[1]}) + {a[2]})"
-    m["tile.gemv"] = _torch_fn("matmul", 2)
-    m["tile.gemv_acc"] = lambda a, _kw: f"({a[0]} + torch.matmul({a[1]}, {a[2]}))"
-    m["tile.gemv_bias"] = lambda a, _kw: f"(torch.matmul({a[0]}, {a[1]}) + {a[2]})"
+    # tile matmul variants — .float() to match hardware FP32 accumulation output
+    m["tile.matmul"] = lambda a, _kw: f"torch.matmul({a[0]}, {a[1]}).float()"
+    m["tile.matmul_acc"] = lambda a, _kw: f"({a[0]} + torch.matmul({a[1]}, {a[2]}).float())"
+    m["tile.matmul_bias"] = lambda a, _kw: f"(torch.matmul({a[0]}, {a[1]}).float() + {a[2]})"
+    m["tile.gemv"] = lambda a, _kw: f"torch.matmul({a[0]}, {a[1]}).float()"
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/debug/torch_codegen.py:304]: Acknowledged — the `.float()` behavior is covered by the existing `test_tile_matmul_acc` unit test (updated assertion) and exercised end-to-end by the system tests. Will add dedicated unit tests for non-accumulate variants as a follow-up.

File: python/pypto/debug/torch_codegen.py [hunk 5]
```diff
@@ -626,24 +698,60 @@ def get_output(self) -> str:
...
+    "visit_bit_and",
+    "visit_bit_or",
+    "visit_bit_xor",
+    "visit_bit_shift_left",
+    "visit_bit_shift_right",
+):
+    setattr(TorchCodegen, _method_name, TorchCodegen.visit_binary_expr)
+
+for _method_name in ("visit_neg", "visit_not", "visit_bit_not", "visit_abs", "visit_cast"):
+    setattr(TorchCodegen, _method_name, TorchCodegen.visit_unary_expr)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/debug/torch_codegen.py:734]: The `setattr` pattern was chosen deliberately to avoid ~70 lines of identical one-line delegate methods. The nanobind trampoline resolves methods through standard Python attribute lookup, so dynamically-set class attributes work correctly. Trade-off acknowledged — keeping as-is for now.

File: python/pypto/debug/torch_codegen.py [hunk 6]
```diff
@@ -63,11 +63,23 @@ def _torch_dtype(dt: DataType) -> str:
...
 def _tile_load(tensor, offsets, shapes):
     slices = tuple(slice(o, o + s) for o, s in zip(offsets, shapes))
-    return tensor[slices].clone()
+    tile = tensor[slices].clone()
+    # Pad to requested shape if source is smaller (mirrors hardware valid_shapes)
+    if tile.shape != tuple(shapes):
+        padded = torch.zeros(shapes, dtype=tile.dtype)
+        pad_slices = tuple(slice(0, s) for s in tile.shape)
+        padded[pad_slices] = tile
+        return padded
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/debug/torch_codegen.py]: Fixed in 3e1dfe46 — changed to `tile.new_zeros(shapes)` which inherits device and dtype from `tile`.

File: python/pypto/debug/torch_codegen.py [hunk 7]
```diff
@@ -171,7 +183,8 @@ def _handler(a: list[str], kw: dict[str, Any]) -> str:
 
 def _handle_slice(a: list[str], _kw: dict[str, Any]) -> str:
     # args: [tensor_or_tile, shape_tuple, offset_tuple, ...]
-    return f"_tile_load({a[0]}, {a[2]}, {a[1]})"
+    # Use view (not clone) so writes propagate back to the original tensor.
+    return f"_tensor_slice({a[0]}, {a[2]}, {a[1]})"
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/debug/torch_codegen.py:189]: Investigated — in the paged attention IR, `tensor.slice` is used on output tensors where writes must propagate back. Using clone breaks end-to-end correctness (output stays all zeros). Both `tensor.slice` and `tile.slice` use view semantics in torch codegen intentionally. Added a clarifying comment.
